rust/src/parse.rs

```diff
@@ -0,0 +1,143 @@
+use log::error;
+
+use crate::VcfError;
+use noodles_vcf::{
+    self as vcf,
+    variant::record::info::{self, field::Value as InfoValue},
+};
+use pyo3::prelude::*;
+
+#[derive(Debug)]
+pub enum VrsVcfField {
+    VrsAlleleIds,
+    VrsStarts,
+    VrsEnds,
+    VrsStates,
+}
+
+impl VrsVcfField {
+    fn as_str(&self) -> &'static str {
+        match self {
+            VrsVcfField::VrsAlleleIds => "VRS_Allele_IDs",
+            VrsVcfField::VrsStarts => "VRS_Starts",
+            VrsVcfField::VrsEnds => "VRS_Ends",
+            VrsVcfField::VrsStates => "VRS_States",
+        }
+    }
+}
+
+/// Extracts an array of strings from a VCF record's info field.
+///
+/// This function attempts to retrieve the info field specified by `field` from the given VCF record info
+/// using the provided `header`. It expects the field to be an array of strings, and if successful,
+/// returns a vector of `String` values. If the field is not found, not an array, or is not stored as strings,
+/// an error is returned.
+///
+/// # Arguments
+///
+/// * `info` - The VCF record info object.
+/// * `header` - A reference to the VCF header used to interpret the info field.
+/// * `field` - The field identifier as a `VrsVcfField`.
+///
+/// # Returns
+///
+/// A `Result` containing a vector of strings if the info field is successfully extracted and converted;
+/// otherwise, returns a `PyErr`.
+///
+/// # Errors
+///
+/// Returns an error if:
+/// - The specified field is not present in the info.
+/// - The field is present but not stored as an array of strings.
+/// - The array variant does not match the expected type.
+pub fn get_str_info_field(
+    info: vcf::record::Info,
+    header: &vcf::Header,
+    field: VrsVcfField,
+) -> Result<Vec<String>, PyErr> {
+    if let Some(Ok(Some(InfoValue::Array(ids_array)))) = info.get(header, field.as_str()) {
```

Comment:
a lil confused with Rust here, where does `ids_array` get defined?

Reply:
basically, this is syntactic sugar for saying that if the output of `info.get` can successfully be parsed, without error, as an INFO value that's an array, then take that array value and define it as the variable `ids_array` for the included scope. (the idea of `Some` is that it designates a non-null value, and `Ok` contains a value that didn't raise an error). 

in this case it's in the context of an `if let` block but this is specific syntax comes from structural pattern matching: https://doc.rust-lang.org/rust-by-example/flow_control/match/destructuring/destructure_enum.html --